File: Core/Workspace/Typescript/libs/workspace/adapters/memory/src/lib/Database/DatabaseState.ts
import { PushRequestRole } from '@allors/protocol/json/system';
import { IObject, ISession } from '@allors/workspace/domain/system';
import { Class, MethodType, RelationType, RoleType } from '@allors/workspace/meta/system';
import { ChangeSet } from '../ChangeSet';
import { Strategy } from '../Strategy';

export class DatabaseState {
  private changedRoleByRelationType: Map<RelationType, any>;

  private previousDatabaseObject: DatabaseObject;
  private previousChangedRoleByRelationType: Map<RelationType, object>;

  constructor(private readonly strategy: Strategy, public databaseObject?: DatabaseObject) {
    this.strategy = strategy;
    this.databaseObject = databaseObject ?? this.database.Get(this.identity);
    this.previousDatabaseObject = this.databaseObject;
  }

  get hasDatabaseChanges(): boolean {
    return !!this.databaseObject || !!this.changedRoleByRelationType;
  }

  get existDatabaseObjects(): boolean {
    return !!this.databaseObject;
  }

  get version(): number {
    return this.databaseObject?.Version ?? 0;
  }

  get identity(): number {
    return this.strategy.id;
  }

  get Class(): Class {
    return this.strategy.class;
  }

  get session(): ISession {
    return this.strategy.session;
  }

  get Database(): Database {
    return this.session.database;
  }

  canRead(roleType: RoleType): boolean {
    if (!this.existDatabaseObjects) {
      return true;
    }

    let permission = this.session.workspace.database.getPermission(this.class, roleType, Operations.Read);
    return this.databaseObject.IsPermitted(permission);
  }

  canWrite(roleType: RoleType): boolean {
    if (!this.existDatabaseObjects) {
      return true;
    }

    let permission = this.session.workspace.database.getPermission(this.Class, roleType, Operations.Write);
    return this.databaseObject.IsPermitted(permission);
  }

  canExecute(methodType: MethodType): boolean {
    if (!this.existDatabaseObjects) {
      return true;
    }

    let permission = this.session.workspace.database.getPermission(this.Class, methodType, Operations.Execute);
    return this.databaseObject.IsPermitted(permission);
  }

  getRole(roleType: RoleType): unknown {
    if (roleType.objectType.isUnit) {
      return this.changedRoleByRelationType?.get(roleType.relationType);
    }

    if (roleType.isOne) {
      if (this.changedRoleByRelationType?.has(roleType.relationType)) {
        return (this.changedRoleByRelationType?.get(roleType.relationType) as Strategy)?.object;
      }

      let identity = this.databaseObject?.getRole(roleType);
      workspaceRole = this.session.getStrategy(identity);

      return workspaceRole?.Object;
    }

    if (this.changedRoleByRelationType?.has(roleType.relationType)) {
      const workspaceRoles = this.changedRoleByRelationType?.get(roleType.relationType) as Strategy[];
      return workspaceRoles != null ? workspaceRoles.map((v) => v.object) : [];
    }

    let identities = this.databaseObject?.getRole(roleType);
    return !!identities ? identities.map((v) => this.session.getMany<IObject>(v)) : [];
  }

  setUnitRole(roleType: RoleType, role: unknown): void {
    let previousRole = this.getRole(roleType);
    if (Equals(previousRole, role)) {
      return;
    }

    this.changedRoleByRelationType ??= new Map();
    this.changedRoleByRelationType.set(roleType.relationType, role);

    this.session.onChange(this);
  }

  setCompositeRole(roleType: RoleType, value: IObject) {
    let role = value;
    let previousRole = this.getRole(roleType);
    if (previousRole === role) {
      return;
    }

    // OneToOne
    if (previousRole != null) {
      let associationType = roleType.associationType;
      if (associationType.isOne) {
        let previousAssociationObject = this.session.getAssociation<IObject>(previousRole.strategy, associationType);
        previousAssociationObject?.Strategy.Set(roleType, null);
      }
    }

    this.changedRoleByRelationType ??= new Map();
    this.changedRoleByRelationType.set(roleType.relationType, role?.strategy);

    this.session.onChange(this);
  }

  setCompositesRole(roleType: RoleType, value: IObject[]): void {
    let previousRole = this.getRole(roleType) as IObject[];

    let role: IObject[] = [];
    if (value != null) {
      role = value;
    }

    let addedRoles = role.filter((v) => !previousRole.includes(v));
    let removedRoles = previousRole.filter((v) => !role.includes(v));

    if (addedRoles.length === 0 && removedRoles.length === 0) {
      return;
    }

    // OneToMany
    if (previousRole.length > 0) {
      let associationType = roleType.associationType;
      if (associationType.isOne) {
        let addedObjects = this.session.getMany<IObject>(addedRoles);
        for (let addedObject of addedObjects) {
          let previousAssociationObject = this.session.getAssociation<IObject>(addedObject.strategy, associationType);
          previousAssociationObject?.Strategy.Remove(roleType, addedObject);
        }
      }
    }

    this.changedRoleByRelationType ??= new Map();
    this.changedRoleByRelationType.set(
      roleType.relationType,
      role.map((v) => v.strategy)
    );

    this.session.onChange(this);
  }

  reset(): void {
    this.databaseObject = this.database.get(this.identity);
    this.changedRoleByRelationType = null;
  }

  checkpoint(changeSet: ChangeSet): void {
    // Same workspace object
    if (this.databaseObject.Version == this.previousDatabaseObject.Version) {
      // No previous changed roles
      if (this.previousChangedRoleByRelationType == null) {
        if (this.changedRoleByRelationType != null) {
          // Changed roles
          for (const kvp of this.changedRoleByRelationType) {
            const [relationType, cooked] = kvp;
            var raw = this.databaseObject.GetRole(relationType.RoleType);

            changeSet.diffCookedWithRaw(this.strategy, relationType, cooked, raw);
          }
        }
      }
      // Previous changed roles
      else {
        for (const kvp in this.changedRoleByRelationType) {
          const [relationType, role] = kvp;

          const previousRole = this.previousChangedRoleByRelationType.get(relationType);
          changeSet.diffCookedWithCooked(this.strategy, relationType, role, previousRole);
        }
      }
    }
    // Different workspace objects
    else {
      let hasPreviousCooked = this.previousChangedRoleByRelationType != null;
      let hasCooked = this.changedRoleByRelationType != null;

      for (const roleType in this.class.workspaceRoleTypes) {
        let relationType = roleType.relationType;

        if (hasPreviousCooked && this.previousChangedRoleByRelationType.has(relationType)) {
          const previousCooked = this.previousChangedRoleByRelationType.get(relationType);
          if (hasCooked && this.changedRoleByRelationType.has(relationType)) {
            const cooked = this.changedRoleByRelationType.get(relationType);
            changeSet.diffCookedWithCooked(this.strategy, relationType, cooked, previousCooked);
          } else {
            var raw = this.databaseObject.GetRole(roleType);
            changeSet.diffRawWithCooked(this.strategy, relationType, raw, previousCooked);
          }
        } else {
          let previousRaw = this.previousDatabaseObject?.GetRole(roleType);
          if (hasCooked && this.changedRoleByRelationType.has(relationType)) {
            const cooked = this.changedRoleByRelationType.get(relationType);
            changeSet.diffCookedWithRaw(this.strategy, relationType, cooked, previousRaw);
          } else {
            var raw = this.databaseObject.GetRole(roleType);
            changeSet.diffRawWithRaw(this.strategy, relationType, raw, previousRaw);
          }
        }
      }
    }

    this.previousDatabaseObject = this.databaseObject;
    this.previousChangedRoleByRelationType = this.changedRoleByRelationType;
  }

  pushResponse(newDatabaseObject: DatabaseObject): void {
    this.databaseObject = newDatabaseObject;
  }

  pushNew(): PushRequestNewObject {
    return {
      WorkspaceId = this.Identity,
      ObjectType = this.Class.Tag,
      Roles = this.pushRoles(),
    };
  }

  pushExisting(): PushRequestObject {
    return {
      DatabaseId = this.identity,
      Version = this.bersion,
      Roles = this.PushRoles(),
    };
  }

  pushRoles(): PushRequestRole[] {
    if (this.changedRoleByRelationType?.length > 0) {
      const roles: PushRequestRole[] = [];

      for (const kvp of this.changedRoleByRelationType) {
        const [relationType, roleValue] = kvp;

        let pushRequestRole: PushRequestRole = { t: relationType.tag };

        if (relationType.roleType.objectType.isUnit) {
          pushRequestRole.u = UnitConvert.toString(roleValue);
        } else {
          if (relationType.roleType.isOne) {
            pushRequestRole.c = roleValue?.Id;
          } else {
            let roleIds = roleValue.map((v) => v.id);
            if (!this.existDatabaseObjects) {
              pushRequestRole.a = roleIds;
            } else {
              let databaseRole = this.databaseObject.getRole(relationType.roleType);
              if (databaseRole == null) {
                pushRequestRole.a = roleIds;
              } else {
                pushRequestRole.a = roleIds.Except(databaseRole).ToArray();
                pushRequestRole.r = databaseRole.Except(roleIds).ToArray();
              }
            }
          }
        }

        roles.push(pushRequestRole);
      }

      return roles;
    }

    return null;
  }

  isAssociationForRole(roleType: RoleType, forRole: Strategy): boolean {
    if (roleType.objectType.isUnit) {
      return false;
    }

    if (roleType.isOne) {
      if (this.changedRoleByRelationType?.has(roleType.relationType)) {
        const workspaceRole = this.changedRoleByRelationType.get(roleType.relationType);
        return workspaceRole?.Equals(forRole) == true;
      }

      const identity = this.databaseObject?.getRole(roleType);
      return identity === forRole.id;
    }

    if (this.changedRoleByRelationType.has(roleType.relationType)) {
      const workspaceRoles = this.changedRoleByRelationType.get(roleType.relationType);
      return workspaceRoles?.Contains(forRole) == true;
    }

    const identities = this.databaseObject?.GetRole(roleType);
    return identities?.includes(forRole.id);
  }

  diff(): RelationType[] {
    if (!this.changedRoleByRelationType) {
      return [];
    }

    return this.changedRoleByRelationType.keys;
  }
}

File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Database/DatabaseState.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json.Api.Push;
    using Allors.Workspace;
    using Remote;
    using Allors.Workspace.Meta;

    internal sealed class DatabaseState
    {
        private readonly Strategy strategy;

        private DatabaseObject databaseObject;
        private Dictionary<IRelationType, object> changedRoleByRelationType;

        private DatabaseObject previousDatabaseObject;
        private Dictionary<IRelationType, object> previousChangedRoleByRelationType;

        internal DatabaseState(Strategy strategy, DatabaseObject databaseObject = null)
        {
            this.strategy = strategy;
            this.databaseObject = databaseObject ?? this.Database.Get(this.Identity);
            this.previousDatabaseObject = this.databaseObject;
        }

        internal bool HasDatabaseChanges => this.databaseObject == null || this.changedRoleByRelationType != null;

        private bool ExistDatabaseObjects => this.databaseObject != null;

        internal long Version => this.databaseObject?.Version ?? 0;

        private long Identity => this.strategy.Id;

        private IClass Class => this.strategy.Class;

        private Session Session => this.strategy.Session;

        private Database Database => this.Session.Database;

        public bool CanRead(IRoleType roleType)
        {
            if (!this.ExistDatabaseObjects)
            {
                return true;
            }

            var permission = this.Session.Workspace.Database.GetPermission(this.Class, roleType, Operations.Read);
            return this.databaseObject.IsPermitted(permission);
        }

        public bool CanWrite(IRoleType roleType)
        {
            if (!this.ExistDatabaseObjects)
            {
                return true;
            }

            var permission = this.Session.Workspace.Database.GetPermission(this.Class, roleType, Operations.Write);
            return this.databaseObject.IsPermitted(permission);
        }

        public bool CanExecute(IMethodType methodType)
        {
            if (!this.ExistDatabaseObjects)
            {
                return true;
            }

            var permission = this.Session.Workspace.Database.GetPermission(this.Class, methodType, Operations.Execute);
            return this.databaseObject.IsPermitted(permission);
        }

        internal object GetRole(IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {

                if (this.changedRoleByRelationType == null || !this.changedRoleByRelationType.TryGetValue(roleType.RelationType, out var unit))
                {
                    unit = this.databaseObject?.GetRole(roleType);
                }

                return unit;
            }

            if (roleType.IsOne)
            {
                if (this.changedRoleByRelationType != null &&
                    this.changedRoleByRelationType.TryGet<Strategy>(roleType.RelationType, out var workspaceRole))
                {
                    return workspaceRole?.Object;
                }

                var identity = (long?)this.databaseObject?.GetRole(roleType);
                workspaceRole = this.Session.GetStrategy(identity);

                return workspaceRole?.Object;
            }

            if (this.changedRoleByRelationType != null &&
                this.changedRoleByRelationType.TryGet<Strategy[]>(roleType.RelationType, out var workspaceRoles))
            {
                return workspaceRoles != null ? workspaceRoles.Select(v => v.Object).ToArray() : Array.Empty<IObject>();
            }

            var identities = (long[])this.databaseObject?.GetRole(roleType);
            return identities == null ? Array.Empty<IObject>() : identities.Select(v => this.Session.Get<IObject>(v)).ToArray();
        }

        internal void SetUnitRole(IRoleType roleType, object role)
        {
            var previousRole = this.GetRole(roleType);
            if (Equals(previousRole, role))
            {
                return;
            }

            this.changedRoleByRelationType ??= new Dictionary<IRelationType, object>();
            this.changedRoleByRelationType[roleType.RelationType] = role;

            this.Session.OnChange(this);
        }

        internal void SetCompositeRole(IRoleType roleType, object value)
        {
            var role = (IObject)value;
            var previousRole = (IObject)this.GetRole(roleType);
            if (Equals(previousRole, role))
            {
                return;
            }

            // OneToOne
            if (previousRole != null)
            {
                var associationType = roleType.AssociationType;
                if (associationType.IsOne)
                {
                    var previousAssociationObject = this.Session.GetAssociation<IObject>((Strategy)previousRole.Strategy, associationType).FirstOrDefault();
                    previousAssociationObject?.Strategy.Set(roleType, null);
                }
            }

            this.changedRoleByRelationType ??= new Dictionary<IRelationType, object>();
            this.changedRoleByRelationType[roleType.RelationType] = role?.Strategy;

            this.Session.OnChange(this);
        }

        internal void SetCompositesRole(IRoleType roleType, object value)
        {
            var previousRole = (IObject[])this.GetRole(roleType);

            var role = Array.Empty<IObject>();
            if (value != null)
            {
                role = ((IEnumerable<IObject>)value).ToArray();
            }

            var addedRoles = role.Except(previousRole).ToArray();
            var removedRoles = previousRole.Except(role).ToArray();

            if (addedRoles.Length == 0 && removedRoles.Length == 0)
            {
                return;
            }

            // OneToMany
            if (previousRole.Length > 0)
            {
                var associationType = roleType.AssociationType;
                if (associationType.IsOne)
                {
                    var addedObjects = this.Session.Get<IObject>(addedRoles);
                    foreach (var addedObject in addedObjects)
                    {
                        var previousAssociationObject = this.Session.GetAssociation<IObject>((Strategy)addedObject.Strategy, associationType).FirstOrDefault();
                        previousAssociationObject?.Strategy.Remove(roleType, addedObject);
                    }
                }
            }

            this.changedRoleByRelationType ??= new Dictionary<IRelationType, object>();
            this.changedRoleByRelationType[roleType.RelationType] = role.Select(v => (Strategy)v.Strategy).ToArray();

            this.Session.OnChange(this);
        }

        internal void Reset()
        {
            this.databaseObject = this.Database.Get(this.Identity);
            this.changedRoleByRelationType = null;
        }

        internal void Checkpoint(ChangeSet changeSet)
        {
            // Same workspace object
            if (this.databaseObject.Version == this.previousDatabaseObject.Version)
            {
                // No previous changed roles
                if (this.previousChangedRoleByRelationType == null)
                {
                    if (this.changedRoleByRelationType != null)
                    {
                        // Changed roles
                        foreach (var kvp in this.changedRoleByRelationType)
                        {
                            var relationType = kvp.Key;
                            var cooked = kvp.Value;
                            var raw = this.databaseObject.GetRole(relationType.RoleType);

                            changeSet.DiffCookedWithRaw(this.strategy, relationType, cooked, raw);
                        }
                    }
                }
                // Previous changed roles
                else
                {
                    foreach (var kvp in this.changedRoleByRelationType)
                    {
                        var relationType = kvp.Key;
                        var role = kvp.Value;

                        _ = this.previousChangedRoleByRelationType.TryGetValue(relationType, out var previousRole);
                        changeSet.DiffCookedWithCooked(this.strategy, relationType, role, previousRole);
                    }
                }
            }
            // Different workspace objects
            else
            {
                var hasPreviousCooked = this.previousChangedRoleByRelationType != null;
                var hasCooked = this.changedRoleByRelationType != null;

                foreach (var roleType in this.Class.WorkspaceRoleTypes)
                {
                    var relationType = roleType.RelationType;

                    if (hasPreviousCooked && this.previousChangedRoleByRelationType.TryGetValue(relationType, out var previousCooked))
                    {
                        if (hasCooked && this.changedRoleByRelationType.TryGetValue(relationType, out var cooked))
                        {
                            changeSet.DiffCookedWithCooked(this.strategy, relationType, cooked, previousCooked);
                        }
                        else
                        {
                            var raw = this.databaseObject.GetRole(roleType);
                            changeSet.DiffRawWithCooked(this.strategy, relationType, raw, previousCooked);
                        }
                    }
                    else
                    {
                        var previousRaw = this.previousDatabaseObject?.GetRole(roleType);
                        if (hasCooked && this.changedRoleByRelationType.TryGetValue(relationType, out var cooked) == true)
                        {
                            changeSet.DiffCookedWithRaw(this.strategy, relationType, cooked, previousRaw);
                        }
                        else
                        {
                            var raw = this.databaseObject.GetRole(roleType);
                            changeSet.DiffRawWithRaw(this.strategy, relationType, raw, previousRaw);
                        }
                    }
                }
            }

            this.previousDatabaseObject = this.databaseObject;
            this.previousChangedRoleByRelationType = this.changedRoleByRelationType;
        }

        internal void PushResponse(DatabaseObject newDatabaseObject) => this.databaseObject = newDatabaseObject;

        internal PushRequestNewObject PushNew() => new PushRequestNewObject
        {
            WorkspaceId = this.Identity,
            ObjectType = this.Class.Tag,
            Roles = this.PushRoles(),
        };

        internal PushRequestObject PushExisting() => new PushRequestObject
        {
            DatabaseId = this.Identity,
            Version = this.Version,
            Roles = this.PushRoles(),
        };

        private PushRequestRole[] PushRoles()
        {
            if (this.changedRoleByRelationType?.Count > 0)
            {
                var roles = new List<PushRequestRole>();

                foreach (var keyValuePair in this.changedRoleByRelationType)
                {
                    var relationType = keyValuePair.Key;
                    var roleValue = keyValuePair.Value;

                    var pushRequestRole = new PushRequestRole { RelationType = relationType.Tag };

                    if (relationType.RoleType.ObjectType.IsUnit)
                    {
                        pushRequestRole.SetUnitRole = UnitConvert.ToJson(roleValue);
                    }
                    else
                    {
                        if (relationType.RoleType.IsOne)
                        {
                            pushRequestRole.SetCompositeRole = ((Strategy)roleValue)?.Id;
                        }
                        else
                        {
                            var roleIds = ((Strategy[])roleValue).Select(v => v.Id).ToArray();
                            if (!this.ExistDatabaseObjects)
                            {
                                pushRequestRole.AddCompositesRole = roleIds;
                            }
                            else
                            {
                                var databaseRole = (long[])this.databaseObject.GetRole(relationType.RoleType);
                                if (databaseRole == null)
                                {
                                    pushRequestRole.AddCompositesRole = roleIds;
                                }
                                else
                                {
                                    pushRequestRole.AddCompositesRole = roleIds.Except(databaseRole).ToArray();
                                    pushRequestRole.RemoveCompositesRole = databaseRole.Except(roleIds).ToArray();
                                }
                            }
                        }
                    }

                    roles.Add(pushRequestRole);
                }

                return roles.ToArray();
            }

            return null;
        }

        public bool IsAssociationForRole(IRoleType roleType, Strategy forRole)
        {
            if (roleType.ObjectType.IsUnit)
            {
                return false;
            }

            if (roleType.IsOne)
            {
                if (this.changedRoleByRelationType != null &&
                    this.changedRoleByRelationType.TryGet<Strategy>(roleType.RelationType, out var workspaceRole))
                {
                    return workspaceRole?.Equals(forRole) == true;
                }

                var identity = (long?)this.databaseObject?.GetRole(roleType);
                return identity?.Equals(forRole.Id) == true;
            }

            if (this.changedRoleByRelationType != null &&
                this.changedRoleByRelationType.TryGet<Strategy[]>(roleType.RelationType, out var workspaceRoles))
            {
                return workspaceRoles?.Contains(forRole) == true;
            }

            var identities = (long[])this.databaseObject?.GetRole(roleType);
            return identities?.Contains(forRole.Id) == true;
        }
    }
}
